docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+metadata:
+  name: infrastructure
+  namespace: shoot--foo--bar
+spec:
+  type: azure
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foo--bar
+  providerConfig:
+    apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureConfig
+    resourceGroup:
+      name: mygroup
+    networks:
+      vnet: # specify either 'name' or 'cidr'
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
+status:
+  state:
+    terraformState: |
+      {
+          "version": 3,
+          "terraform_version": "0.11.14",
+          "serial": 2,
+          "lineage": "3a1e2faa-e7b6-f5f0-5043-368dd8ea6c10",
+          "modules": [
+              {
+              }
+          ]
+          ...
+      }
+```
+
+
+### Shoot reconciliation flow changes
+
+The only data which must be stored in the `ShootState` by the Garden Controller Manager are secrets (e.g ca for the apiserver). Therefore the `botanist.DeploySecrets` step is changed. It is split into two functions which take a list of secrets that have to be generated.
+- `botanist.GenerateSecretState` Generates certificate authorities and other secrets which have to be persisted in the ShootState and must not be regenerated on the New Seed.
+- `botanist.DeploySecrets` Takes secret data from the `ShootState`, generates new ones (e.g. client tls certificates from the saved certificate authorities) and deploys everything in the Shoot's control plane on the New Seed
+
+
+### ShootState synchronization controller
+
+The ShootState synchronization controller is part of the garden controller manager. It syncs the state of extension custom resources from the shoot namespace to the garden cluster and updates the correspodnign `ShootState` resource's `spec.extensions` field. The sync happens every `X` minutes and always runs before migration starts (if possible).
+
+
+### Migration workflow
+1. Starting migration
+    - Migration can only be started after a Shoot cluster has been successfully created so that the `status.seed` field in the `Shoot` resource has been set
+    - The `Shoot` resource's field `spec.seedName="new-seed"` is edited to hold the name of the _New Seed_ and reconciliation is automatically triggered
+    - The Garden Controller Manager checks if the equality between `spec.seedName` and `status.seed`, detects that they are different and triggers migration.
+1. The Garden Controller Manager waits for the _New Seed_ to be ready
+1. ShootState is synced from the _Host Seed_ cluster
+    - The Garden Controller Manager lists all deployed extension resources for the Shoot and copies the values in their `status.state` field to the `ShootState` resource
+    - If the _Host Seed's_ Apiserver is not available, then last successfull synced `ShotoState` is used
+1. Stop the Shoot's api server
+1. Scale Down the Shoot's control plane in the _Host Seed_. ETCD backup will be done automatically before the ETCD pod is scaled down.
```

Comment:
That may change slightly with the ETCD druid (by @georgekuruvillak), but can be sorted out also later.